compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -465,15 +465,17 @@ where
 
     prof_timer.finish_with_query_invocation_id(dep_node_index.into());
 
-    let diagnostics = diagnostics.into_inner();
-    let side_effects = QuerySideEffects { diagnostics };
+    let side_effects = QuerySideEffects { diagnostics: diagnostics.into_inner() };
 
     if std::intrinsics::unlikely(!side_effects.is_empty()) {
         if Q::ANON {
             qcx.store_side_effects_for_anon_node(dep_node_index, side_effects);
         } else {
             qcx.store_side_effects(dep_node_index, side_effects);
         }
+    } else {
+        // Avoid generating code for the destructor here
+        mem::forget(side_effects);
```

Comment:
This needs a comment to explain why it's not a memory leak.

Reply:
Added.